enhancements/network/network-policy-audit-logging.md

```diff
@@ -0,0 +1,357 @@
+ // defaults to 20 messages per second
+ // +optional
+ aclLoggingRateLimit  *uint32 `json:"aclLoggingRateLimit,omitempty"`
+ <--- END NEW CODE --->
+
+}
+```
+
+The flag `aclLoggingRateLimit` defaults to `20` messages per second.
+
+This will result in an updated OVNKubernetes Operator configuration object: `(.spec.defaultNetwork.ovnKubernetesConfig)` which will enable cluster-wide configuration of ACL logging at cluster installation time:
+
+```go
+spec:
+  defaultNetwork:
+    type: OVNKubernetes
+    ovnKubernetesConfig:
+      mtu: 1400
+      genevePort: 6081
+      ipsecConfig: {}
+      aclLoggingRateLimit: 20
+```
+
+To enable the ACL logging on a per namespace basis using metadata annotations in the namespace's definition like the following
+
+```go
+kind: Namespace
+apiVersion: v1
+metadata:
+  name: tenantA
+  annotations:
+    k8s.ovn.org/acl-logging: '{ "deny": "alert", "allow": "notice" }'
+```
+
+or
+
+```go
+kind: Namespace
+apiVersion: v1
+metadata:
+  name: tenantB
+  annotations:
+    k8s.ovn.org/acl-logging: '{ "deny": "notice" }'
+```
+
+The logging can be activated for either allow, drop, or allow and drop actions. The severity must be one of `alert`, `warning`,
+`notice`, `info`, or `debug` as described in [OVN documentation](http://www.openvswitch.org/support/dist-docs/ovn-nbctl.8.html).
+
+each level is listed in order or descending severity
+
+```go
+alert    A major failure forced a process to abort.
+
+warning  A high-level operation or a subsystem failed.  Attention
+         is warranted.
+
+notice   A low-level operation failed, but higher-level subsystems
+         may be able to recover.
+
+info     Information that may be useful in retrospect when
+         investigating a problem.
+
+debug    Information useful only to someone with intricate
+         knowledge of the system, or that would commonly cause too-
+         voluminous log output.  Log messages at this level are not
+         logged by default.
+```
+
+### User Stories
+
+#### Story 1
+
+As an Openshift user I want to see what Network Policies and accompanied ACLs are dropping the most traffic in a cluster. For example, when a packet runs into a `drop` ACL it will log a message similar to the following:
+
+ ```go
+  2021-01-05T17:34:02.675Z|00004|acl_log(ovn_pinctrl0)|INFO|name="<Network Policy Name>", verdict=drop, severity=info: icmp,vlan_tci=0x0000,dl_src=50:54:00:00:00:02,dl_dst=50:54:00:00:00:01,nw_src=192.168.0.3,nw_dst=192.168.0.2,nw_tos=0,nw_ecn=0,nw_ttl=64,icmp_type=8,icmp_code=0
+ ```
+
+#### Story 2
+
+As an Openshift user I want to control the rate of ACL logging to ensure I can extract the useful information involving accept/reject actions in high hit rate scenarios.
+
+#### Story 3
+
+As an Openshift user I want to monitor in realtime traffic that is hitting any network policies I have implemented.
+
+--- TBD in implementation (Realtime Monitoring would be nice here)
+
+### Implementation Details/Notes/Constraints
```

Comment:
might need work in OVN to get the log solmewhere else?  or work in ovn-kube to get the data somewhere else ...

Reply:
I am working to get the ACL logging into it's own file

Reply:
any tracker for this you could link to?  then we could resolve it as an open item

Reply:
So after a week messing around with trying to use OVS's `syslog` mechanism + a sidecar container I think we will just be able to use native OVN to export only the ACL logs to a file using the `ovn-controller's `--logfile flag` (Since currently OCP does not even use that feature) I.e we shouldn't need any OVN changes